sdk/inc/azure/core/internal/az_hfsm_pipeline_internal.h

```diff
@@ -0,0 +1,180 @@
+/* Copyright (c) Microsoft Corporation. All rights reserved. */
+/* SPDX-License-Identifier: MIT */
+
+/**
+ * @file az_hfsm_pipeline.h
+ * @brief Definition of #az_hfsm_pipeline and related types describing a bi-directional HFSM
+ *        pipeline.
+ *
+ * @remarks Both a non-blocking I/O (default) as well as a blocking I/O implementations are
+ * available. To enable the blocking mode, set TRANSPORT_MQTT_SYNC.
+ *
+ * @note The blocking I/O model (when TRANSPORT_MQTT_SYNC is defined) is not thread-safe.
+ */
+
+#ifndef _az_HFSM_PIPELINE_H
+#define _az_HFSM_PIPELINE_H
+
+// Define AZ_HFSM_PIPELINE_SYNC to manually control when pipeline events are executed.
+// The default action is to immediately process events on the same stack.
+// #define AZ_HFSM_PIPELINE_SYNC
+
+#include <azure/core/internal/az_hfsm_internal.h>
+#include <azure/core/az_platform.h>
+#include <azure/core/az_result.h>
+#include <stdint.h>
+
+#include <azure/core/_az_cfg_prefix.h>
+
+/**
+ * @brief Used to declare HFSM policies.
+ *
+ */
+
+typedef struct az_hfsm_policy az_hfsm_policy;
+
+typedef struct az_hfsm_pipeline az_hfsm_pipeline;
+
+/**
+ * @brief The type representing a HFSM with dispatch capabilities. Derived from #az_hfsm.
+ *
+ */
+struct az_hfsm_policy
+{
+  az_hfsm hfsm; // Must be the first element to properly cast the struct to az_hfsm.
+  az_hfsm_policy* inbound;
+  az_hfsm_policy* outbound;
+  az_hfsm_pipeline* pipeline;
+};
+
+/**
+ * @brief Internal definition of an MQTT pipeline.
+ *
+ */
+struct az_hfsm_pipeline
+{
+  struct
+  {
+    az_hfsm_policy* outbound_handler;
+    az_hfsm_policy* inbound_handler;
+#ifndef TRANSPORT_MQTT_SYNC
+    az_platform_mutex mutex;
+#endif
+  } _internal;
+};
+
+AZ_NODISCARD az_result az_hfsm_pipeline_init(
```

Comment:
Missing documentation here.

Please ensure that all other functions have the proper documentation too (I won't leave other comments re this.)